src/Common/Fiber.h

```diff
@@ -4,4 +4,181 @@
 #include <base/defines.h>
 #include <boost/context/fiber.hpp>
 
-using Fiber = boost::context::fiber;
+/// Class wrapper for boost::context::fiber.
+/// It tracks current executing fiber for thread and
+/// supports storing fiber-specific data
+/// that will be destroyed on fiber destructor.
+class Fiber
+{
+public:
+    struct CleanUpFn
+    {
+        virtual void operator()(void *) = 0;
+        virtual ~CleanUpFn() = default;
+    };
+
+    using Impl = boost::context::fiber;
+
+    template< typename StackAlloc, typename Fn>
+    Fiber(StackAlloc && salloc, Fn && fn) : impl(std::allocator_arg_t(), std::forward<StackAlloc>(salloc), std::forward<Fn>(fn))
+    {
+    }
+
+    Fiber() = default;
+
+    Fiber(Fiber && other) = default;
+    Fiber & operator=(Fiber && other) = default;
+
+    Fiber(const Fiber &) = delete;
+    Fiber & operator =(const Fiber &) = delete;
+
+    ~Fiber()
+    {
+        for (auto & [_, data] : local_data)
+            (*data.cleanup_fn)(data.ptr);
+    }
+
+    explicit operator bool() const
+    {
+        return impl.operator bool();
+    }
+
+    void resume()
+    {
+        /// Update information about current executing fiber.
+        auto & current_fiber_info = getCurrentFiberInfo();
+        auto parent_fiber_info = current_fiber_info;
+        current_fiber_info = FiberInfo{this, &parent_fiber_info};
+        impl = std::move(impl).resume();
+        /// Restore current fiber info.
+        current_fiber_info = parent_fiber_info;
+    }
+
+    /// Set pointer to fiber-specific data, it will be stored in hash-map
+    /// using provided key and cleaned on fiber destructor using provided
+    /// cleanup function.
+    void setLocalData(void * key, void * ptr, CleanUpFn * cleanup_fn)
```

Comment:
Need a comment what's key here.

Also probably make `setLocalData/getLocalData` private and available only from FiberLocal